arches/app/models/migrations/11613_nodegroup_root_node_constraints.py

```diff
@@ -0,0 +1,33 @@
+# Generated by Django 5.1.3 on 2024-11-19 09:33
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("models", "11613_nodegroup_root_node"),
+    ]
+
+    operations = [
+        migrations.AddConstraint(
+            model_name="node",
+            constraint=models.CheckConstraint(
+                condition=models.Q(
+                    ("istopnode", True), ("nodegroup__isnull", False), _connector="OR"
+                ),
+                name="has_nodegroup_or_istopnode",
```

Comment:
I though nodes _always_ had a nodegroup? And the difference was that if a node is a top node its nodegroup shares an ID with it.

Reply:
Top nodes don't, for better or worse!

Reply:
I will double-check tomorrow morning, but I'm relatively sure they do? This was an issue with promoting/demoting nodes into their own nodegroups in the editable future graphs that eventually pushed me to do it all in JSON. Please let me double-check this before merging!

Reply:
There's a lil [diagram](https://arches.readthedocs.io/en/stable/developing/reference/data-model/#resource-model-overview) showing a free floating top node ☁️

Reply:
oh fun! I just checked on my local and see that it makes an extra nodegroup. We should touch base early tomorrow to verify each other's steps and see where the disconnect is 🔌

Reply:
> Please let me double-check this before merging!

of course, no rush by any means! enjoy your time off!

Reply:
whoops yeah, just verified and you're right 👍